utils/visualizer/qviz/content_loader.py

```diff
@@ -2,250 +2,261 @@
+
+    if revision_key in metadata.keys():
+            metadata_key_str = metadata[revision_key]
+            metadata_key = json.loads(metadata_key_str)
+            dimension_count = len(metadata_key["columnTransformers"])
+            symbol_count = (dimension_count + 5) // 6
+            print(symbol_count)
+            return metadata_key, symbol_count
+    else:
+        print(f"No metadata found for the given RevisionID. There is no Qbeast table.")
+        return None
 
-def load_single_log_file(file_path: str, revision_id: str) -> (list[dict], set[str]):
+# 3. Returns a dictionary, the keys are the cubeids and the values are the actual cubes
+def extract_cubes_from_blocks(delta_table: DeltaTable, symbol_count: int) -> dict:
     """
-    Extract valid AddFiles from a single json log file. An AddFile is valid
-    if there's no RemoveFile with the same path.
-    example log file entries:
-    {"metaData":{"id":"c109cfb6-a770-4c02-b004-d44bbe91e981", ...}}
-    {"add":{"path":"b4542891-5b03-40cc-8ef1-293493e21814.parquet", ...}}
-    {"remove":{"path":"272bbd79-2bf0-444d-b15e-8e4326c9d281.parquet", ...}}
-    :param file_path: json log file path
-    :param revision_id: target RevisionID
-    :return: a list of AddFiles, and a set of RemoveFile paths
+    Takes all the blocks stored in the table and creates the cubes.
+    :param delta_table: list of json log files
+    :param symbol_count: Integer representing the number of symbols used for the representation
+    :return: A dictionary with all the cubes stored in the table.
     """
-    with open(file_path, 'r') as f:
-        lines = f.readlines()
-
-    add_files = list()
-    remove_paths = set()
-    for line in lines:
-        action_file = json.loads(line)
-        # AddFile entry
-        if 'add' in action_file:
-            add_file = action_file['add']
-            _id = add_file['tags']['revision']
-            if _id == revision_id:
-                # Add revision AddFile to result
-                add_files.append(add_file)
-        # RemoveFile entry
-        elif 'remove' in action_file:
-            remove_file = action_file['remove']
-            remove_paths.add(remove_file['path'])
-
-    return add_files, remove_paths
-
-
-def extract_revision_metadata(checkpoint_file: str, json_files: list[str], revision_id: str) -> dict:
+
+    df = delta_table.get_add_actions(True).to_pandas()
+
+    df_filtered = df[["size_bytes", "tags.blocks"]]
+    # In this dictionary we store a cube for every cubeId. Each cube will be associated to his cube ID
+    cubes_dict = dict()
+
+    # Iterate throughout each row of the dataframe
+    for index, row in df_filtered.iterrows():
+        blocks_str = row[
+            "tags.blocks"
+        ]  # This returns a pandas series with the blocks of the cube in each of the iterated rows
+        # Check if the cube is a JSON chain & try to convert it into a list of dictionaries
+        if isinstance(blocks_str, str):
```

Comment:
You should do something if the `blocks_str` is not a string. Could you try to find a scenario where this is the case?

Reply:
I've checked with three different tables and in all of them blocks_str is a string... maybe because of how the delta table is defined it will always be a string.

Reply:
About doing something if blocks_str is not a string... to avoid repeating code I will simply make sure before entering the if that it is a string

Reply:
yep, let's just complete the if else 😄